# Universidad del Norte
## IST 4330 - Estructuras Discretas
### Luis Cárcamo Morales

### Invariante

#### Ejemplo 1

<code>def f(n):
  s = 0
  k = 0
  while k <= n:
    s = s + k        
    k = k + 1
    
  return s</code>

In [20]:
def f(n):
    s = 0
    k = 0
    while k <= n:
        s = s + k
        print("En la iteración k = %d, s = %d" % (k ,s))
        assert s == k*(k+1)//2 # DEBUG
        k = k + 1
        
    assert s == n*(n+1)//2
    return s

In [21]:
f(10)

En la iteración k = 0, s = 0
En la iteración k = 1, s = 1
En la iteración k = 2, s = 3
En la iteración k = 3, s = 6
En la iteración k = 4, s = 10
En la iteración k = 5, s = 15
En la iteración k = 6, s = 21
En la iteración k = 7, s = 28
En la iteración k = 8, s = 36
En la iteración k = 9, s = 45
En la iteración k = 10, s = 55


55

<table>
    <tr>
        <td>
            $\textbf{k}$
        </td>
        <td>
            0
        </td>
        <td>
            1
        </td>
        <td>
            2
        </td>
        <td>
            3
        </td>
        <td>
            4
        </td>
        <td>
            5
        </td>
        <td>
            $\cdots$
        </td>
    </tr>
    <tr>
        <td>
            $\textbf{s}$
        </td>
        <td>
            0
        </td>
        <td>
            1
        </td>
        <td>
            3
        </td>
        <td>
            6
        </td>
        <td>
            10
        </td>
        <td>
            15
        </td>
        <td>
            $\cdots$
        </td>
    </tr>
</table>

##### Invariante
En la $k$-ésima iteración, el valor de $s$ es:

$$s_k = \sum_{i=0}^{k}{i} = \cfrac{k\cdot (k+1)}{2}$$

##### Poscondición
Al salir del ciclo repetitivo, el valor de $s$ es:

$$s = \sum_{0\leq k \leq n}{k}=\cfrac{1}{2}\cdot n\cdot (n+1)$$

### Ejemplo 2

<code>def g(n):
  s = 0
  k = 1
  while k <= n:
    s = s * k
    k = k + 1
    
  return s</code>

In [25]:
def g(n):
    s = 1
    k = 1
    while k <= n:
        s *= k
        k += 1
    
    return s

In [26]:
g(5)

120

##### Invariante
En la $k$-ésima iteración, el valor de $s$ es:

$$s_k = \prod_{i=1}^{k}{i} = k!$$

##### Poscondición
Al salir del ciclo repetitivo, el valor de $s$ es:

$$s = \prod_{1\leq k \leq n}{k}=n!$$

<table>
    <tr>
        <td>
            $\textbf{k}$
        </td>
        <td>
            1
        </td>
        <td>
            2
        </td>
        <td>
            3
        </td>
        <td>
            4
        </td>
        <td>
            5
        </td>
        <td>
            $\cdots$
        </td>
    </tr>
    <tr>
        <td>
            $\textbf{s}$
        </td>
        <td>
            1
        </td>
        <td>
            2
        </td>
        <td>
            6
        </td>
        <td>
            24
        </td>
        <td>
            120
        </td>
        <td>
            $\cdots$
        </td>
    </tr>
</table>

### Soluciones Enteras

In [29]:
from scipy.special import comb

#### Ejemplo 1

\begin{align*}
x_1 + x_2 + x_3 + x_4 + x_5 + x_6&= 15\\
\textit{sujeto a:}\quad& 0\leq x_i \leq 9,\quad i=1,2,...,5
\end{align*}

In [28]:
s = 0
for x1 in range(0, 10):
    for x2 in range(0, 10):
        for x3 in range(0, 10):
            for x4 in range(0, 10):
                for x5 in range(0, 10):
                    x6 = 15 - x1 - x2 - x3 - x4 - x5
                    if x6>=0 and x6 <=9:
                        s += 1
                        
print(s)

13992


In [32]:
fn1 = comb(6+15-1,15)-comb(6,1)*comb(6+15-10-1,15-10)
print(fn1)

13992.0


#### Ejemplo 2

$$x1+x2+x3+x4=17.$$
Sujeto a $$x1\geq0,\,x2\geq1,\,x3\geq2,\,x4\geq3$$

In [33]:
s = 0
for x2 in range(1, 18):
    for x3 in range(2, 18):
        for x4 in range(3, 18):
            x1 = 17 - x2 - x3 - x4
            if x1 >=0:
                s += 1
                
print(s)

364


In [34]:
fn1 = comb(4+17-6-1,17-6)
print(fn1)

364.0


#### Ejemplo 3

\begin{equation*}
x_1+x_2+x_3=15
\end{equation*}
sujeto a 
\begin{align*}
0\leq x_1\leq10,\,x_2=3,\,x_3\geq1
\end{align*}


In [35]:
s = 0
x2 = 3
for x1 in range(0, 11):
    x3 = 15-x1-x2
    if x3 >= 1:
        s+=1

print(s)

11


In [39]:
fn2 = comb(2+11-1,11)-comb(2+0-1,0)
print(fn2)

11.0
